src/api/tap.rs

```diff
@@ -0,0 +1,465 @@
+use super::{ControlMessage, ControlSender};
+use crate::{
+    config::ConfigDiff,
+    event::{Event, LogEvent},
+    topology::{fanout, RunningTopology},
+};
+use futures::{channel::mpsc as futures_mpsc, SinkExt, StreamExt};
+use itertools::Itertools;
+use parking_lot::RwLock;
+use std::{
+    cmp::Ordering,
+    collections::HashSet,
+    fmt::Debug,
+    hash::{Hash, Hasher},
+    sync::{Arc, Weak},
+};
+use tokio::sync::mpsc as tokio_mpsc;
+use uuid::Uuid;
+use weak_table::WeakHashSet;
+
+type TapSender = tokio_mpsc::Sender<TapResult>;
+
+/// Clients can supply glob patterns to find matched topology components.
```

Comment:
Just to clarify, is this the same logic that the topology config also allows for wildcard `inputs`? I think it'd be good to make sure they match, maybe even pulling out into a separate shared lib to ensure they don't drift.

Reply:
No, tap currently uses the `glob` lib -- akin to how we're using it in file sources.

What we have for input expansion is:

https://github.com/timberio/vector/blob/8297828fa1e9d6a83c45f60e0942303611c19764/src/config/compiler.rs#L89-L120

Effectively, this expands only where `*` is the last character. It wouldn't match on all glob patterns, such as `?` or `*` appearing in the middle of a string.

This is probably the majority use case and we may not need the fuller regex pattern. I'm happy to move this to a separate crate if we prefer parity.

Reply:
I think parity here would be nice. I would have expected the matching to be the same between the input here and using wildcards for component names in the config file.

Reply:
Just confirming here that I retained wildcard inputs for now, pending the outcome of https://github.com/timberio/vector/issues/6786.